### ISSUE TITLE
Monitoring is crippling the cluster (very high load avg), scraping at 3200 metrics/second

### ISSUE THREAD

**Issue by riaan53** [rancher/rancher#19248]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
1. Create 3 node cluster via RKE (weave CNI, RancherOS 1.5.1)
2. Install Rancher HA via helm
3. Note `top` stats and other metrics on nodes
4. Enable monitoring (use default settings except 2GB ram for prom)
5. Note `top` stats and other metrics on nodes and Grafana metrics
6. Add prometheus datasource stats dashboard to Grafana to get more insight

**Result:**
Excessive resource usage noticed, looks like the culprit is kube-apiserver as seen in graphs below.

Note the load averages.

Without monitoring:
![rancher-without-monitoring](https://user-images.githubusercontent.com/7716342/55164377-36441b00-5174-11e9-8eca-c2be07da9558.png)

With monitoring:
![Screenshot 2019-03-28 at 15 55 48](https://user-images.githubusercontent.com/7716342/55164421-4c51db80-5174-11e9-85fb-92f41cbae4b9.png)
![Screenshot 2019-03-28 at 15 56 08](https://user-images.githubusercontent.com/7716342/55164432-5378e980-5174-11e9-9f04-06380a917050.png)

![Screenshot 2019-03-28 at 16 21 17](https://user-images.githubusercontent.com/7716342/55165172-94253280-5175-11e9-944f-b10459fe517e.png)

Metrics scraped (prometheus stats dashboard):
![Screenshot 2019-03-28 at 16 14 36](https://user-images.githubusercontent.com/7716342/55164611-9dfa6600-5174-11e9-9ee4-0615f9729069.png)

![image (1)](https://user-images.githubusercontent.com/7716342/55165232-bcad2c80-5175-11e9-9964-a27ec32499d6.png)

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): Rancher 2.2.0
- Installation option (single install/HA): HA
- Rancher OS 1.5.1

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): RKE (0.2.0) create and HA Rancher (2.2.0) installed vir Helm on Digitalocean 
- Machine type (cloud/VM/metal) and specifications (CPU/memory): Digitalocean, 8GB ram, 4 x vCPU.
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T22:29:25Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:30:26Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
[rancher@rancheros-s-4vcpu-8gb-fra1-03 ~]$ docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:20:43 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:28:38 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Comment by riaan53** [comment]
Prev one was on RancherOS 1.5.1 tried it now on Ubuntu 18.04.2 with same results. Used exactly the same VMs specs and versions as above on DigitalOceon. 

Before monitoring enabled:
<img width="1336" alt="Screenshot 2019-03-28 at 21 25 37" src="https://user-images.githubusercontent.com/7716342/55187256-2d1d7300-51a1-11e9-8a05-95b7a1035d29.png">


After:
<img width="1310" alt="Screenshot 2019-03-28 at 21 34 08" src="https://user-images.githubusercontent.com/7716342/55187286-46262400-51a1-11e9-8b19-f8baaab7d863.png">

<img width="1564" alt="Screenshot 2019-03-28 at 21 34 58" src="https://user-images.githubusercontent.com/7716342/55187321-6655e300-51a1-11e9-8973-9f1a57f502d4.png">
<img width="1335" alt="Screenshot 2019-03-28 at 21 35 11" src="https://user-images.githubusercontent.com/7716342/55187341-71107800-51a1-11e9-9e98-90ba9796fdbb.png">

RKE cluster creation file
```
# If you intened to deploy Kubernetes in an air-gapped environment,
# please consult the documentation on how to configure custom RKE images.
nodes:
- address: 138.68.66.190
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: ""
  user: root
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
- address: 138.68.68.226
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: ""
  user: root
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
- address: 139.59.153.8
  port: "22"
  internal_address: ""
  role:
  - controlplane
  - worker
  - etcd
  hostname_override: ""
  user: root
  docker_socket: /var/run/docker.sock
  ssh_key: ""
  ssh_key_path: ~/.ssh/id_rsa
  ssh_cert: ""
  ssh_cert_path: ""
  labels: {}
services:
  etcd:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    external_urls: []
    ca_cert: ""
    cert: ""
    key: ""
    path: ""
    snapshot: null
    retention: ""
    creation: ""
    backup_config: null
  kube-api:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    service_cluster_ip_range: 10.43.0.0/16
    service_node_port_range: ""
    pod_security_policy: false
    always_pull_images: false
  kube-controller:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    cluster_cidr: 10.42.0.0/16
    service_cluster_ip_range: 10.43.0.0/16
  scheduler:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
  kubelet:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
    cluster_domain: cluster.local
    infra_container_image: ""
    cluster_dns_server: 10.43.0.10
    fail_swap_on: false
  kubeproxy:
    image: ""
    extra_args: {}
    extra_binds: []
    extra_env: []
network:
    plugin: weave
    options:
      WeavePassword: somepassword
authentication:
  strategy: x509
  sans: []
  webhook: null
addons: ""
addons_include: []
system_images:
  etcd: rancher/coreos-etcd:v3.2.24-rancher1
  alpine: rancher/rke-tools:v0.1.27
  nginx_proxy: rancher/rke-tools:v0.1.27
  cert_downloader: rancher/rke-tools:v0.1.27
  kubernetes_services_sidecar: rancher/rke-tools:v0.1.27
  kubedns: rancher/k8s-dns-kube-dns:1.15.0
  dnsmasq: rancher/k8s-dns-dnsmasq-nanny:1.15.0
  kubedns_sidecar: rancher/k8s-dns-sidecar:1.15.0
  kubedns_autoscaler: rancher/cluster-proportional-autoscaler:1.0.0
  coredns: coredns/coredns:1.2.6
  coredns_autoscaler: rancher/cluster-proportional-autoscaler:1.0.0
  kubernetes: rancher/hyperkube:v1.13.4-rancher1
  flannel: rancher/coreos-flannel:v0.10.0-rancher1
  flannel_cni: rancher/flannel-cni:v0.3.0-rancher1
  calico_node: rancher/calico-node:v3.4.0
  calico_cni: rancher/calico-cni:v3.4.0
  calico_controllers: ""
  calico_ctl: rancher/calico-ctl:v2.0.0
  canal_node: rancher/calico-node:v3.4.0
  canal_cni: rancher/calico-cni:v3.4.0
  canal_flannel: rancher/coreos-flannel:v0.10.0
  weave_node: weaveworks/weave-kube:2.5.0
  weave_cni: weaveworks/weave-npc:2.5.0
  pod_infra_container: rancher/pause:3.1
  ingress: rancher/nginx-ingress-controller:0.21.0-rancher3
  ingress_backend: rancher/nginx-ingress-controller-defaultbackend:1.4-rancher1
  metrics_server: rancher/metrics-server:v0.3.1
ssh_key_path: ~/.ssh/id_rsa
ssh_cert_path: ""
ssh_agent_auth: false
authorization:
  mode: rbac
  options: {}
ignore_docker_version: false
kubernetes_version: ""
private_registries: []
ingress:
  provider: ""
  options: {}
  node_selector: {}
  extra_args: {}
cluster_name: ""
cloud_provider:
  name: ""
prefix_path: ""
addon_job_timeout: 0
bastion_host:
  address: ""
  port: ""
  user: ""
  ssh_key: ""
  ssh_key_path: ""
  ssh_cert: ""
  ssh_cert_path: ""
monitoring:
  provider: ""
  options: {}
restore:
  restore: false
  snapshot_name: ""
dns: null
```

**Comment by hpdvanwyk** [comment]
I am also seeing this. It only seems to happen on clusters created with the standalone rke binary. Using Rancher and adding a custom cluster creates one with sane load.

The most obvious difference is the disk IO. One two single node clusters, one created with rke and one with Rancher:

rke disk io:
![rke disk io](https://user-images.githubusercontent.com/948289/55229509-e45bcd80-5225-11e9-97fb-2670286e888c.png)

Rancher disk io:
![rancher disk io](https://user-images.githubusercontent.com/948289/55229530-f8073400-5225-11e9-81fa-3c67c5d4b768.png)

Iotop with rke on the right and Rancher on the left:
![rke vs rancher](https://user-images.githubusercontent.com/948289/55229558-0e14f480-5226-11e9-9bd3-7b2b3f96579a.png)

It looks like etcd is doing considerably more writes on the cluster created with rke if monitoring is turned on.

**Comment by thxCode** [comment]
@riaan53 , this is a bug, we will fix it in next release. This will be reproduced only in `local` cluster, thank you for the detail report, cc @hpdvanwyk. 

The high CPU load causes by an error resource modification race between `server` and `cluster-agent`. 

Prometheus ingests 3200 metrics/s is fine, on next, we will do some optimization, like reducing the scraping rate for Kubernetes components, reusing record rules, and so on. Meanwhile, we are going to introduce a dashboard https://github.com/rancher/rancher/issues/19261 to help you watch Prometheus.

**Comment by riaan53** [comment]
Hi @thxCode , Thank you for fixing! I have done a quick test on 2.2.2RC and it seems to do the trick. Also looking forward to the new prom dashboard on by default :)

**Comment by jiaqiluo** [comment]
The bug is reproduced in Rancher v2.2.1
Setup: 
- Rancher is installed on 3 nodes; each node is an AWS EC2 t2.xlarge instance(4 vCPU, 16 GB Memory)
- Besides of the `local` cluster, I also add an EC2 cluster(3 nodes, t2.medium) an imported RKE cluster (3 nodes, t2.medium)

Steps for each cluster:
- use the `top` command to track the CPU usage of the process running `kube-apiserver` on each node before enabling cluster-level monitoring
- enable cluster-level monitoring
- keep tracking how the CPU usage changes

Result:
- in the `local` cluster, it observes that the `kube-apiserver` uses a significant amount of CPU after enabling cluster-level monitoring. the change is roughly from 4% ~ 5% to   20% ~ 60% 

The CPU usage of the process running `kube-apiserver` before and after enabling cluster-level monitoring 

Cluster Type | before | after
------------ | ------------- | -------------
EC2 cluster | 2 - 3% | 3 - 6% 
Imported RKE cluster | 2 - 3% | 5 - 8%
Local cluster | 4% ~ 5% | 20 - 60%
--------------

The bug fix is validated in Rancher v2.2.2-rc1 by using the same configuration and repeating the same steps to test in Rancher v2.2.2-rc1. 

Result:
- All three kinds of clusters are using a reasonable amount of CPU resource before and after enabling cluster-level monitoring.  
- For the `local` cluster, the CPU usage changes from 4% - 5% to 5% - 8%

The CPU usage of the process running `kube-apiserver` before and after enabling cluster-level monitoring 

Cluster Type | before | after
------------ | ------------- | -------------
EC2 cluster | 6 - 8% | 4 - 8%
Imported RKE cluster | 1 - 2% |  1 - 4%
Local cluster |4% - 21% |   4 - 28%

The CPU usage of the `local` cluster after enabling the cluster-level monitoring
<img width="1333" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/55513805-7ba49480-561b-11e9-9840-7aa6ca651611.png">

**Comment by jiaqiluo** [comment]
This issue needs to be validated on Rancher: master again for the milestone v2.3

**Comment by jiaqiluo** [comment]
The bug fix is validated in Rancher: master  7725bc8  4/16

Setup: 
- Rancher is installed on 3 nodes; each node is an AWS EC2 t2.xlarge instance(4 vCPU, 16 GB Memory)
- Besides of the `local` cluster, I also add an imported RKE cluster (3 nodes, t2.medium)

Steps for each cluster:
- use the `top` command to track the CPU usage of the process running `kube-apiserver` on each node before enabling cluster-level monitoring
- enable cluster-level monitoring
- keep tracking how the CPU usage changes

Result:

The CPU usage of the process running `kube-apiserver` before and after enabling cluster-level monitoring 

Cluster Type | before | after
------------ | ------------- | ------------- 
Imported RKE cluster | 1 - 2% |  1 - 5%
Local cluster |2% - 16% |   4 - 11%
